llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -136,13 +140,95 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorShiftsByZero(
   return true;
 }
 
+bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  MachineInstr *ExtOp = MRI.getUniqueVRegDef(MI.getOperand(2).getReg());
+
+  assert(ExtOp && "Expected SSA.");
+  if (ExtOp->getOpcode() != AIE2::G_INTRINSIC ||
+      cast<GIntrinsic>(*ExtOp).getIntrinsicID() !=
+          Intrinsic::aie2_extract_I128_I512)
+    return false;
+
+  MachineIRBuilder MIRBuilder(MI);
+  MIRBuilder.buildCopy(DstReg, ExtOp->getOperand(2).getReg());
+  ExtOp->eraseFromParent();
```

Comment:
Probably here we may corrupt the iterator., do we need to erase ExtOp? DCE should handle this case.